matscipy/cauchy_born.py

```diff
@@ -0,0 +1,742 @@
+        atoms_copy.set_cell(scaled_cell,scale_atoms=True)
+        f_after = atoms_copy.get_forces()
+
+        force_diff = f_after-f_before
+
+        #transform force_diff to lattice frame
+        force_diff_lattice = np.zeros_like(force_diff)
+        for i in range(len(atoms)):
+            force_diff_lattice[i,:] = A@force_diff[i,:]
+
+        #generate the lattice masks, checking to ensure that if by chance no shift occurs in one direction due to the applied strains 
+        #that the other directions are used instead and things don't break
+        lattice1mask = force_diff_lattice[:,0]>0
+        lattice2mask = force_diff_lattice[:,0]<0
+        if all(element == lattice1mask[0] for element in lattice1mask): #if they're all false or true, try the y component
+            lattice1mask = force_diff_lattice[:,1]>0
+            lattice2mask = force_diff_lattice[:,1]<0
+            if all(element == lattice1mask[0] for element in lattice1mask): #if they're all false, try the z component
+                lattice1mask = force_diff_lattice[:,2]>0
+                lattice2mask = force_diff_lattice[:,2]<0
+        
+
+        #set the masks for which atoms are in each of the sublattices
+        self.lattice1mask = lattice1mask
+        self.lattice2mask = lattice2mask     
+
+        
+
+    
+
+    def fit_taylor(self,de=1e-4):
+        '''fit a simple taylor-expansion type model to predict the cauchy-born shift corrector at a given applied strain
+            using finite differences. Sets self.grad_f to the vector of first derivatives of the first shift component 
+            with each of the 6 strain components and self.hess_f to the hessian of the first shift component with the strain components.
+            NOTE this model is generally worse than the model provided later in initial_lasso_fit.
+        Parameters
+        ----------
+        de : float
+            tolerance for finite difference gradient calculations
+        '''
+        calc = self.calc
+        def f_gl(E_vec,calc,unitcell):
+            '''Function that calculates the first component of the cauchy-born shift corrector for a given green-lagrange strain tensor.
+            Parameters
+            ----------
+            E_vec : array_like
+                green-lagrange strain state to apply in voigt notation
+            calc : ASE calculator object
+                ASE calculator to use to find shift
+            unitcell : ASE atoms object
+                Single unit cell of crystal to apply strain vector to and to find shift
+
+            Returns
+            -------
+            shift_diff[0] : float
+                first component of cauchy-born shift correction for applied strain state
+            '''
+            #green lagrange version of function
+            #turn E into matrix
+            E = np.zeros([3,3])
+            E[0,0] = E_vec[0]
+            E[1,1] = E_vec[1]
+            E[2,2] = E_vec[2]
+            E[2,1],E[1,2] = E_vec[3],E_vec[3]
```

Comment:
pep8: comma should be followed by space (later occurances not noted, but please fix them all!)